x/ibc/light-clients/07-tendermint/types/proposal_handle.go

```diff
@@ -12,116 +10,120 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/ibc/core/exported"
 )
 
-// CheckProposedHeaderAndUpdateState will try to update the client with the new header if and
-// only if the proposal passes and one of the following two conditions is satisfied:
-// 		1) AllowUpdateAfterExpiry=true and Expire(ctx.BlockTime) = true
-// 		2) AllowUpdateAfterMisbehaviour and IsFrozen() = true
-// In case 2) before trying to update the client, the client will be unfrozen by resetting
-// the FrozenHeight to the zero Height. If AllowUpdateAfterMisbehaviour is set to true,
-// expired clients will also be updated even if AllowUpdateAfterExpiry is set to false.
-// Note, that even if the update happens, it may not be successful. The header may fail
-// validation checks and an error will be returned in that case.
-func (cs ClientState) CheckProposedHeaderAndUpdateState(
-	ctx sdk.Context, cdc codec.BinaryMarshaler, clientStore sdk.KVStore,
-	header exported.Header,
-) (exported.ClientState, exported.ConsensusState, error) {
-	tmHeader, ok := header.(*Header)
+// CheckProposedHeaderAndUpdateState will try to update the client with the state of the
+// substitute if and only if the proposal passes and one of the following conditions  are
+// satisfied:
+//		1) The substitute client is the same type as the subject client
+//		2) The subject and substitute client states match in all parameters (expect frozen height, latest height, and chain-id)
+// 		3) AllowUpdateAfterMisbehaviour and IsFrozen() = true
+// 		4) AllowUpdateAfterExpiry=true and Expire(ctx.BlockTime) = true
+// In case 3) before updating the client, the client will be unfrozen by resetting
+// the FrozenHeight to the zero Height. If a client is frozen and AllowUpdateAfterMisbehaviour
+// is set to true, the client will be unexpired even if AllowUpdateAfterExpiry is set to false.
+// Note, that even if the subject is updated to the state of the substitute, an error may be
+// returned if the updated client state is invalid or the client is expired.
+func (cs ClientState) CheckSubstituteAndUpdateState(
+	ctx sdk.Context, cdc codec.BinaryMarshaler, subjectClientStore,
+	substituteClientStore sdk.KVStore, substituteClient exported.ClientState,
+	initialHeight exported.Height,
+) (exported.ClientState, error) {
+	substituteClientState, ok := substituteClient.(*ClientState)
 	if !ok {
-		return nil, nil, sdkerrors.Wrapf(
-			clienttypes.ErrInvalidHeader, "expected type %T, got %T", &Header{}, header,
+		return nil, sdkerrors.Wrapf(
+			clienttypes.ErrInvalidClient, "expected type %T, got %T", &ClientState{}, substituteClient,
 		)
 	}
 
+	// substitute clients are not allowed to be upgraded during the voting period
+	if substituteClientState.GetLatestHeight().GetRevisionNumber() != initialHeight.GetRevisionNumber() {
```

Comment:
NOTE: This logic may cause neighboring consensus states to have a time difference of more than the subject's trusting period.

I personally don't believe that is an issue.